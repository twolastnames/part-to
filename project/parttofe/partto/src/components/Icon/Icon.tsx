import React, { ForwardRefExoticComponent, RefAttributes } from "react";
import { rem } from "@mantine/core";
import {
  Icon as TablerIcon,
  IconProps as TablerIconProps,
  IconPlayerTrackNext,
  IconPlayerPlay,
  IconList,
  IconFile,
  IconCaretDown,
  IconCaretUp,
  IconChefHat,
  IconCooker,
  IconToolsKitchen,
} from "@tabler/icons-react";

import classes from "./Icon.module.scss";

export type IconType = ForwardRefExoticComponent<
  TablerIconProps & RefAttributes<TablerIcon>
>;

export enum Size {
  Small = 20,
  Medium = 30,
  Large = 50,
  ExtraLarge = 80,
}

export interface IconProps {
  definition: IconType;
  size?: Size;
  onClick?: () => void;
}

export function Icon({ onClick, definition, size }: IconProps) {
  return React.createElement(definition, {
    ...(onClick
      ? {
          onClick,
          className: classes.clickable,
        }
      : {}),
    style: {
      width: rem((size || Size.Medium) as number),
      height: rem((size || Size.Medium) as number),
    },
  });
}

export const Next = IconPlayerTrackNext;
export const Start = IconPlayerPlay;
export const List = IconList;
export const File = IconFile;
export const Up = IconCaretUp;
export const Down = IconCaretDown;
export const ChefHat = IconChefHat;
export const Oven = IconCooker;
export const KitchenTools = IconToolsKitchen;
