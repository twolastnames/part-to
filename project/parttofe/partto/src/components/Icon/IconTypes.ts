import {
  Icon as TablerIcon,
  IconProps as TablerIconProps,
} from "@tabler/icons-react";
import { ForwardRefExoticComponent, RefAttributes } from "react";

export type IconType = ForwardRefExoticComponent<
  TablerIconProps & RefAttributes<TablerIcon>
>;

export enum Size {
  Small = 20,
  Medium = 30,
  Large = 50,
  ExtraLarge = 80,
}

export interface IconProps {
  definition: IconType;
  size?: Size;
  onClick?: () => void;
}
