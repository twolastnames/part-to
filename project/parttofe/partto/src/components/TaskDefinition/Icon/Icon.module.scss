@import "../../../App.scss";

.icon {
  padding: var(--gap-medium);
}

.hidden {
  display: none;
}
