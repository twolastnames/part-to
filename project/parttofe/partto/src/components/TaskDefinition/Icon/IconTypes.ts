export type ClassNames = {
  imminent: string;
  duty: string;
  task: string;
};

export interface IconProps {
  classNames: ClassNames;
}
