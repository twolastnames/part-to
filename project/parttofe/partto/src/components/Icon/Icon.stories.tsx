import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Icon, Next } from "./Icon";
import { ShellProvider } from "../../providers/ShellProvider";
import { Size } from "./IconTypes";

const meta: Meta<typeof Icon> = {
  component: Icon,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Icon>;

export const Medium: Story = {
  args: { definition: Next },
};

export const Small: Story = {
  args: { definition: Next, size: Size.Small },
};

export const Large: Story = {
  args: { definition: Next, size: Size.Large },
};

export const ExtraLarge: Story = {
  args: { definition: Next, size: Size.ExtraLarge },
};
