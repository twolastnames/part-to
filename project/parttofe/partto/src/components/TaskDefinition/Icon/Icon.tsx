import React from "react";

import classes from "./Icon.module.scss";
import { ClassNames, IconProps } from "./IconTypes";
import {
  Duty as DutyIcon,
  Imminent as ImminentIcon,
  Icon as PhysicalIcon,
  Task as TaskIcon,
} from "../../Icon/Icon";

export const Duty: ClassNames = {
  imminent: classes.hidden,
  duty: classes.icon,
  task: classes.hidden,
};

export const Task: ClassNames = {
  imminent: classes.hidden,
  duty: classes.hidden,
  task: classes.icon,
};

export const Imminent: ClassNames = {
  imminent: classes.icon,
  duty: classes.hidden,
  task: classes.hidden,
};

export function Icon({ classNames: { task, duty, imminent } }: IconProps) {
  return (
    <>
      <span className={imminent}>
        <PhysicalIcon definition={ImminentIcon} />
      </span>
      <span className={duty}>
        <PhysicalIcon definition={DutyIcon} />
      </span>
      <span className={task}>
        <PhysicalIcon definition={TaskIcon} />
      </span>
    </>
  );
}
