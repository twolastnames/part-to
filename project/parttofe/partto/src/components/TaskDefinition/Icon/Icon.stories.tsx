import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Icon } from "./Icon";
import { ShellProvider } from "../../../providers/ShellProvider";

const meta: Meta<typeof Icon> = {
  component: Icon,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Icon>;

export const Simple: Story = {
  args: {},
};
