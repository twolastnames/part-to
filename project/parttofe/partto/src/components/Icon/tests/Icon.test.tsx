import React from "react";
import { describe, expect, test } from "@jest/globals";
import renderer from "react-test-renderer";
import { Icon, Next } from "../Icon";

describe("Icon", () => {
  test("snapshot", () => {
    expect(
      renderer.create(<Icon definition={Next} />).toJSON(),
    ).toMatchSnapshot();
  });
});
