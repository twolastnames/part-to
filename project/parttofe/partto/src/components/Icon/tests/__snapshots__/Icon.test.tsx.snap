// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon snapshot 1`] = `
<svg
  className="tabler-icon tabler-icon-player-track-next "
  fill="none"
  height={24}
  stroke="currentColor"
  strokeLinecap="round"
  strokeLinejoin="round"
  strokeWidth={2}
  style={
    Object {
      "height": "calc(1.875rem * var(--mantine-scale))",
      "width": "calc(1.875rem * var(--mantine-scale))",
    }
  }
  viewBox="0 0 24 24"
  width={24}
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M3 5v14l8 -7z"
  />
  <path
    d="M14 5v14l8 -7z"
  />
</svg>
`;
